x-pack/plugins/alerts/server/invalidate_pending_api_keys/task.ts

```diff
@@ -0,0 +1,157 @@
+import { InvalidatePendingApiKey } from '../types';
+
+const TASK_TYPE = 'alerts_invalidate_api_keys';
+export const TASK_ID = `Alerts-${TASK_TYPE}`;
+
+const invalidateAPIKey = async (
+  params: InvalidateAPIKeyParams,
+  securityPluginSetup?: SecurityPluginSetup
+): Promise<InvalidateAPIKeyResult> => {
+  if (!securityPluginSetup) {
+    return { apiKeysEnabled: false };
+  }
+  const invalidateAPIKeyResult = await securityPluginSetup.authc.invalidateAPIKeyAsInternalUser(
+    params
+  );
+  // Null when Elasticsearch security is disabled
+  if (!invalidateAPIKeyResult) {
+    return { apiKeysEnabled: false };
+  }
+  return {
+    apiKeysEnabled: true,
+    result: invalidateAPIKeyResult,
+  };
+};
+
+export function initializeAlertsInvalidateApiKeys(
+  logger: Logger,
+  internalSavedObjectsRepository: Promise<{
+    createInternalRepository: () => ISavedObjectsRepository;
+  }>,
+  taskManager: TaskManagerSetupContract,
+  securityPluginSetup?: SecurityPluginSetup
+) {
+  registerAlertsInvalidateApiKeysTask(
+    logger,
+    internalSavedObjectsRepository,
+    taskManager,
+    securityPluginSetup
+  );
+}
+
+export function scheduleAlertsInvalidateApiKeys(
+  logger: Logger,
+  config: Promise<AlertsConfig>,
+  taskManager?: TaskManagerStartContract
+) {
+  if (taskManager) {
+    scheduleTasks(logger, taskManager, config);
+  }
+}
+
+function registerAlertsInvalidateApiKeysTask(
+  logger: Logger,
+  internalSavedObjectsRepository: Promise<{
+    createInternalRepository: () => ISavedObjectsRepository;
+  }>,
+  taskManager: TaskManagerSetupContract,
+  securityPluginSetup?: SecurityPluginSetup
+) {
+  taskManager.registerTaskDefinitions({
+    [TASK_TYPE]: {
+      title: 'Invalidate Alerts API Keys',
+      createTaskRunner: taskRunner(logger, internalSavedObjectsRepository, securityPluginSetup),
+    },
+  });
+}
+
+async function scheduleTasks(
+  logger: Logger,
+  taskManager: TaskManagerStartContract,
+  config: Promise<AlertsConfig>
+) {
+  const interval = (await config).invalidateApiKeysTask.interval;
+  try {
+    await taskManager.ensureScheduled({
+      id: TASK_ID,
+      taskType: TASK_TYPE,
+      schedule: {
+        interval,
+      },
+      state: {},
+      params: {},
+    });
+  } catch (e) {
+    logger.debug(`Error scheduling task, received ${e.message}`);
+  }
+}
+
+function taskRunner(
+  logger: Logger,
+  internalSavedObjectsRepository: Promise<{
+    createInternalRepository: () => ISavedObjectsRepository;
+  }>,
+  securityPluginSetup?: SecurityPluginSetup
+) {
+  return ({ taskInstance }: RunContext) => {
+    const { state } = taskInstance;
+    return {
+      async run() {
+        let totalInvalidated = 0;
+        try {
+          const repository = (await internalSavedObjectsRepository).createInternalRepository();
+          const apiKeysToInvalidate = await repository.find<InvalidatePendingApiKey>({
+            type: 'invalidate_pending_api_key',
+          });
```

Comment:
I wonder if we should filter records that are older than `5m` to allow tasks to use the API key for at least 5m before we invalidate it. I'm also wondering if we should make it configurable, I know SIEM was mentioning they have long running tasks 🤔